### PR TITLE
add support for bbmap read mapper

### PR DESCRIPTION
**Pull request by mmokrejs**

### CODE REVIEW

**Comment by mmokrejs** [comment]
I copied to a tempdir all previously created results from a k=128 run while abyss' map mapper was used. Which of the files should I delete? Addition of `--dry-run align=bbmap` to abyss-pe shows the new script is called but I do not want to waste too much CPU cycles so I am better asking what are the files which should be re-created in this step. Thank you.

**Comment by benvvalk** [comment]
Thanks for your PR, @mmokrejs.

Sorry about the CI check failure -- it has nothing to do with your PR and is something I need to fix.

In general, our criteria for merging PR into ABySS is:
1. Does it improve the results, resource usage, or usability of the assembler?
2. Does it provide a feature that is interest to multiple users?

With respect to 1), would you be willing to compare assembly results on a small data set with abyss-map vs. bwa vs. BBmap?

With respect to 2, do you think there are users that would be interested in using BBmap?

If you don't want to do the testing and just want to use the BBmap patch for your usage, that is totally fine of course.

File: bin/abyss-bbmap [hunk 1]
```diff
@@ -0,0 +1,69 @@
+#!/apps/gentoo/bin/bash
```
Comment (benvvalk) [inline bin/abyss-bbmap]: Please change to something that is portable.  Any one of `#!/usr/bin/bash`, `#!bash`, `#!/usr/bin/env bash` are acceptable in my opinion.

File: bin/abyss-bbmap [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+See https://sourceforge.net/projects/bbmap for more details.
+EOF
+		exit
+		;;
+	--version)
+		cat <<EOF
+abyss-bbmap (ABySS)
+Written by Martin Mokrejs.
+EOF
+		bwa 2>&1 |head -n4
```
Comment (benvvalk) [inline bin/abyss-bbmap]: Please replace this with the command to generate the version info for BBmap

**Comment by benvvalk** [comment]
@mmokrejs To re-generate results with BBmap, remove all files newer than `-3.fa` and `-3.dot`.

**Comment by mmokrejs** [comment]
> To re-generate results with BBmap, remove all files newer than -3.fa and -3.dot.

OK, so here is what has been left:

```
-rw-rw----+  1 mmokrejs mmokrejs     414257 Feb  6 11:35 coverage.hist
-rw-rw----+  1 mmokrejs mmokrejs 2369956822 Feb  6 13:44 tt_16D1C3L12-1.fa
-rw-rw----+  1 mmokrejs mmokrejs  151702095 Feb  6 13:44 tt_16D1C3L12-bubbles.fa
-rw-rw----+  1 mmokrejs mmokrejs  987873363 Feb  6 13:47 tt_16D1C3L12-1.dot
-rw-rw----+  1 mmokrejs mmokrejs          0 Feb  6 13:47 tt_16D1C3L12-1.path
-rw-rw----+  1 mmokrejs mmokrejs  710045063 Feb  6 13:52 tt_16D1C3L12-2.dot1
-rw-rw----+  1 mmokrejs mmokrejs 1794342387 Feb  6 13:53 tt_16D1C3L12-2.fa
-rw-rw----+  1 mmokrejs mmokrejs  710045063 Feb  6 13:53 tt_16D1C3L12-2.dot
-rw-rw----+  1 mmokrejs mmokrejs  671269287 Feb  6 13:57 tt_16D1C3L12-3.dot
-rw-rw----+  1 mmokrejs mmokrejs    4022696 Feb  6 13:57 tt_16D1C3L12-2.path
-rw-rw----+  1 mmokrejs mmokrejs 1726012111 Feb  6 13:58 tt_16D1C3L12-3.fa
lrwxrwxrwx   1 mmokrejs mmokrejs         17 Feb  6 13:58 tt_16D1C3L12-unitigs.fa -> tt_16D1C3L12-3.fa
-rw-rw----+  1 mmokrejs mmokrejs   38509396 Feb  6 13:58 tt_16D1C3L12-indel.fa
```

**Comment by mmokrejs** [comment]
> Does it improve the results, resource usage, or usability of the assembler?

In my experience the insert-sizes of the mate-pairs seem terribly wide and suspicious. Once one asks for only alignments with identity percentage above 96% the distribution is closer to expectations (around 4kb and not insert-sizes of hundreds of kbp, etc.). I do not have a chart handy for bwa but for bbmap without `minid=0.96 idfilter=0.96` the default `minid=0.76 idfilter=0` kicks in and the results are not nice. The calculation for `minid` does not account for gaps so it does not really work for all cases as expected. Therefore, the `idfilter` must be included does the real job. Once `minid=0.96 idfilter=0.96` are added the distributions is not spread up to astronomical insert-sizes. I have drafts of images here:

insert-sizes__bbmap_with_defaults
histogram__bbmap_with_defaults

insert-sizes__bbmap_with_idfilter_96pct
 histogram__bbmap_with_idfilter_96pct

In the abyss-bbmap wrapper I therefore ask bbmap to record insert-sizes and identities in the SAM stream for each mapped read. That was the source of values for the above charts. yes, some values are still unremoved, seems something is still escaping the `idfilter`.


> Does it provide a feature that is interest to multiple users?

This filtering and possibility to discard matches altogether instead of picking up randomly any of them or just the first one is IMHO critical for scaffolding (avoids conflicts in contig links?). See the help text:

```
ambiguous=best 
    (ambig) Set behavior on ambiguously-mapped reads (with 
                 multiple top-scoring mapping locations).
    best    (use the first best site)
    toss    (consider unmapped)
    random  (select one top-scoring site randomly)
    all     (retain all top-scoring sites)
```
I do not know if any other tools provides such filters or not. But I am fine promoting bbmap as it scales well and does its job well done.

> With respect to 1), would you be willing to compare assembly results on a small data set with abyss-map vs. bwa vs. BBmap?

Sure, send it my way.

> With respect to 2, do you think there are users that would be interested in using BBmap?

There are many existing users, it is amazingly fast, scaling well, give it a try. You could drop your k-mer splitting and counting code and let bbmaps's tools do it, and just use output files. Look for `kmercountexact.sh` and other tools in the zip folder.

**Comment by mmokrejs** [comment]
I also use commonly `idhist=$prefix.idhist.txt` and `ihist=$prefix.ihist.txt` to obtain histogram values right away from bbmap. But I did not add that to the abyss-bbmap wrapper as I had troubles to figure how to obtain the `$prefix`. Initially I was studying the arrays in input files but provided the data are passed in via STDIN it was a waste of time. Some tools from BBmap bundle support multiple files like `in1=file,file in2=file,file2,null` if I remember right. Somebody willing to pursue the ability to pass multiple input filenames from abyss-pe down to bbmap.sh should check the documentation of the bbmap.sh script directly. _... maybe it is a feature of just tadpole.sh and not bbmap.sh?_

**Comment by mmokrejs** [comment]
Unfortunately something in the pipeline choked:

```
abyss-bbmap -v  -j21 -l40    /scratch/mygenome/HKMHTBCXX.1.tt_16D1C3L12.trimmomatic.paired.prinseq.minlen20.19552.19552.fastq tt_16D1C3L12__abyss_128-3.fa \
        |abyss-fixmate -v  -l40  -h HKMHTBCXX.1.pe-3.hist \
        |sort -snk3 -k4 \
        |samtools view -Sb - -o HKMHTBCXX.1.pe-3.bam
Reading from standard input...
abyss-tofastq /scratch/mygenome/HKMHTBCXX.1.tt_16D1C3L12.trimmomatic.paired.prinseq.minlen20.19552.19552.fastq | bbmap.sh threads=21 in=stdin.fastq out=stdout.sam ref=tt_16D1C3L12__abyss_128-3.fa
java -Djava.library.path=/scratch/mygenome/BBmap-37.90/jni/ -ea -Xmx37367m -cp /scratch/mygenome/BBmap-37.90/current/ align2.BBMap build=1 overwrite=true fastareadlen=500 threads=21 ambiguous=toss interleaved=t local=t minid=0.96 idfilter=0.96 inserttag=t idtag=t in=stdin.fastq out=stdout.sam ref=tt_16D1C3L12__abyss_128-3.fa
Executing align2.BBMap [build=1, overwrite=true, fastareadlen=500, threads=21, ambiguous=toss, interleaved=t, local=t, minid=0.96, idfilter=0.96, inserttag=t, idtag=t, in=stdin.fastq, out=stdout.sam, ref=tt_16D1C3L12__abyss_128-3.fa]
Version 37.90 [build=1, overwrite=true, fastareadlen=500, threads=21, ambiguous=toss, interleaved=t, local=t, minid=0.96, idfilter=0.96, inserttag=t, idtag=t, in=stdin.fastq, out=stdout.sam, ref=tt_16D1C3L12__abyss_128-3.fa]

Processing reads in paired-ended mode.
Started read stream.
Started 21 mapping threads.
error: `*       0       0       *       *       0       0       TTCAATATTACCTATAACAATTGTTAAATTCTCAAAATGTCACAAGTGTGAGCGAACCTGGCACAAGTGTGAGCGAACCAAACTCATCTCAAGACACATCTCATAACACAAAAACAAATGGATTGGTGCATGGTCTGCAGTTTCAACAATACTGCGACCCCACCCAAAACCAGGAACCCTCTCCAGCAGACCTGAGATAAAATTCTCAGAAGGTGACATGACAATAATACATGTTACCACCACCATTAAAGGGAAACTGAG   DDDDDIIIIIIIIIIIIIIIIIIHIIIIIIIIIIIIIIIIIIIIIIIIIIHIIIIIIIIIIIIIIIIHIIIHIIIDHHIIIIIIIIIIIIIIIIIIIGIIIEHIHHIGIHIIIIIIIIIIIIIIIIIIHHIIHEHIIIIIGHIIIIIHIIIIEHIIHGHIIIIIIICHEFHIHGHIIGHHHHHHIHHFHHIEHGHIGHIFHGIH?HHHHECHGFAHEHIEHI?HHHE?..8FEHIIGEHHHFHCF8-A@GGH@-6FHE-6-'
```

**Comment by benvvalk** [comment]
@mmokrejs By default ABySS programs omit some SAM fields (saves disk space when using the `pe-sam`/`mp-sam` targets), and it is possible that `BBmap` doesn't like that.

Adding `ssq=1` to your `abyss-pe` command will tell ABySS to include all the fields in the SAM output of the various programs and will hopefully fix your issue.

**Comment by mmokrejs** [comment]
Weird, it did not help.

```
abyss-bbmap -v  -j21 -l40    /scratch/mygenome/HKMHTBCXX.1.tt_16D1C3L12.trimmomatic.paired.prinseq.minlen20.19552.19552.fastq tt_16D1C3L12__abyss_128-3.fa \
	|abyss-fixmate-ssq -v  -l40  -h HKMHTBCXX.1.pe-3.hist \
	|sort -snk3 -k4 \
	|samtools view -Sb - -o HKMHTBCXX.1.pe-3.bam
...
error: `*	0	0	*	*	0	0	TTCAATATTACCTATAACAATTGTTAAATTCTCAAAATGTCACAAGTGTGAGCGAACCTGGCACAAGTGTGAGCGAACCAAACTCATCTCAAGACACATCTCATAACACAAAAACAAATGGATTGGTGCATGGTCTGCAGTTTCAACAATACTGCGACCCCACCCAAAACCAGGAACCCTCTCCAGCAGACCTGAGATAAAATTCTCAGAAGGTGACATGACAATAATACATGTTACCACCACCATTAAAGGGAAACTGAG	DDDDDIIIIIIIIIIIIIIIIIIHIIIIIIIIIIIIIIIIIIIIIIIIIIHIIIIIIIIIIIIIIIIHIIIHIIIDHHIIIIIIIIIIIIIIIIIIIGIIIEHIHHIGIHIIIIIIIIIIIIIIIIIIHHIIHEHIIIIIGHIIIIIHIIIIEHIIHGHIIIIIIICHEFHIHGHIIGHHHHHHIHHFHHIEHGHIGHIFHGIH?HHHHECHGFAHEHIEHI?HHHE?..8FEHIIGEHHHFHCF8-A@GGH@-6FHE-6-'
```
I would assume samtools dies becasue bbmap is the one producing correct SAM:

`D3FCO8P1:285:HKMHTBCXX:1:1101:3796:2163	65	77	*	0	0	*	*	0	0	TTCAATATTACCTATAACAATTGTTAAATTCTCAAAATGTCACAAGTGTGAGCGAACCTGGCACAAGTGTGAGCGAACCAAACTCATCTCAAGACACATCTCATAACACAAAAACAAATGGATTGGTGCATGGTCTGCAGTTTCAACAATACTGCGACCCCACCCAAAACCAGGAACCCTCTCCAGCAGACCTGAGATAAAATTCTCAGAAGGTGACATGACAATAATACATGTTACCACCACCATTAAAGGGAAACTGAG	DDDDDIIIIIIIIIIIIIIIIIIHIIIIIIIIIIIIIIIIIIIIIIIIIIHIIIIIIIIIIIIIIIIHIIIHIIIDHHIIIIIIIIIIIIIIIIIIIGIIIEHIHHIGIHIIIIIIIIIIIIIIIIIIHHIIHEHIIIIIGHIIIIIHIIIIEHIIHGHIIIIIIICHEFHIHGHIIGHHHHHHIHHFHHIEHGHIGHIFHGIH?HHHHECHGFAHEHIEHI?HHHE?..8FEHIIGEHHHFHCF8-A@GGH@-6FHE-6-`

**Comment by benvvalk** [comment]
@mmokrejs Hrm. Not sure what is going on.  Can you confirm which program is generating the error by building up the pipeline one command at a time?

It isn't a very helpful error message unfortunately.

**Comment by benvvalk** [comment]
Regarding our earlier discussion about small test datasets, you can download some small datasets with mate pairs here: http://gage.cbcb.umd.edu/data/

**Comment by mmokrejs** [comment]
So both `abyss-fixmate -v  -l40  -h HKMHTBCXX.1.pe-3.hist` and `abyss-fixmate-ssq -v  -l40  -h HKMHTBCXX.1.pe-3.hist` fail with this error. samtools is not involved.

**Comment by benvvalk** [comment]
@mmokrejs I looked at the `abyss-fixmate` code a bit.  Based on where the error message is printed, it looks like it is having a problem parsing the line immediately preceding the SAM line shown in the error message:
```
*	0	0	*	*	0	0	TTCAATATTACCTATAACAATTGTTAAATTCTCAAAATGTCACAAGTGTGAGCGAACCTGGCACAAGTGTGAGCGAACCAAACTCATCTCAAGACACATCTCATAACACAAAAACAAATGGATTGGTGCATGGTCTGCAGTTTCAACAATACTGCGACCCCACCCAAAACCAGGAACCCTCTCCAGCAGACCTGAGATAAAATTCTCAGAAGGTGACATGACAATAATACATGTTACCACCACCATTAAAGGGAAACTGAG	DDDDDIIIIIIIIIIIIIIIIIIHIIIIIIIIIIIIIIIIIIIIIIIIIIHIIIIIIIIIIIIIIIIHIIIHIIIDHHIIIIIIIIIIIIIIIIIIIGIIIEHIHHIGIHIIIIIIIIIIIIIIIIIIHHIIHEHIIIIIGHIIIIIHIIIIEHIIHGHIIIIIIICHEFHIHGHIIGHHHHHHIHHFHHIEHGHIGHIFHGIH?HHHHECHGFAHEHIEHI?HHHE?..8FEHIIGEHHHFHCF8-A@GGH@-6FHE-6-
```

Can you find the line *before* the line above in the BBmap output, and see what it looks like?  I can't imagine what the problem would be, unfortunately, but it seems there is something it can't handle in the BBmap SAM output.

**Comment by mmokrejs** [comment]
It is `@PG` line causing the error.

```
@PG     ID:BBMap        PN:BBMap        VN:37.90        CL:java -Djava.library.path=/scratch/mygenome/BBmap-37.90/jni/ -ea -Xmx37383m align2.BBMap build=1 overwrite=true fastareadlen=500 threads=21 ambiguous=toss interleaved=t lo
cal=t minid=0.96 idfilter=0.96 inserttag=t idtag=t in=stdin.fastq out=stdout.sam ref=tt_16D1C3L12__abyss_128-3.fa
D3FCO8P1:285:HKMHTBCXX:1:1101:3796:2163 65      77      *       0       0       *       *       0       0       TTCAATATTACCTATAACAATTGTTAAATTCTCAAAATGTCACAAGTGTGAGCGAACCTGGCACAAGTGTGAGCGAACCAAACTCATCTCAAGACACATCTCATAACACAAAAACAAATGGATTGGTGCATGGTCTGCAGTTTCAACAATACTGCGACCCCACCCAAAACCAGGAACCCTCTCCAGCAGACCTGAGATAAAATTCTCAGAAGGTGACATGACAATAATACATGTTACCACCACCATTAAAGGGAAACTGAG   DDDDDIIIIIIIIIIIIIIIIIIHIIIIIIIIIIIIIIIIIIIIIIIIIIHIIIIIIIIIIIIIIIIHIIIHIIIDHHIIIIIIIIIIIIIIIIIIIGIIIEHIHHIGIHIIIIIIIIIIIIIIIIIIHHIIHEHIIIIIGHIIIIIHIIIIEHIIHGHIIIIIIICHEFHIHGHIIGHHHHHHIHHFHHIEHGHIGHIFHGIH?HHHHECHGFAHEHIEHI?HHHE?..8FEHIIGEHHHFHCF8-A@GGH@-6FHE-6-
```

**Pull request closed without merging**